### PR TITLE
Properly handle titlecase characters in cow_to_uppercase and cow_to_lowercase

### LINKED ISSUE #2: Discrepancy with `String::to_lowercase()
```rust
use cow_utils::*;

fn main() {
    let s = "ᾈ";
    assert_eq!(s.to_lowercase(), s.cow_to_lowercase())
}
```

```
$ cargo run                 
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
     Running `target/debug/cwl`
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `"ᾀ"`,
 right: `"ᾈ"`', src/main.rs:5:5
```

We ran across this doing some proptesting against some of our code that uses `cow_to_lowercase()`.  `ᾈ` has no particular meaning to us -- it's just the reduced value that `proptest` found.

I see this crate hasn't been updated in about 4 years, but I thought this was useful to report.  We really like `cow-utils`, so even if this is deemed unworthy of fixing, thanks for a great piece of software.

### PR DESCRIPTION
**Pull request by thomcc**
Fixes #2

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -237,3 +239,11 @@ impl<'s> CowUtils<'s> for &'s str {
         }
     }
 }
+
+fn changes_when_lowercased(c: char) -> bool {
+    !core::iter::once(c).eq(c.to_lowercase())
```
Comment (RReverser) [inline src/lib.rs:244]: Why the indirection with single-item iterator instead of just `c == c.to_lowercase()`?

**Review by RReverser — APPROVED** [review]
Thanks!

**Comment by RReverser** [comment]
I just added benchmarks for `to_lowercase` on master and it's pretty unfortunate that even for the best-case scenario of no-op `to_lowercase` Cow version is now 3x slower than regular one whereas w/o this change it's 2x faster:

```
To Lowercase/to_lowercase/ax40
                        time:   [146.32 ns 149.49 ns 152.90 ns]
                        change: [-5.7087% -2.7402% +0.0303%] (p = 0.08 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
  8 (8.00%) high mild
  1 (1.00%) high severe
To Lowercase/cow_to_lowercase/ax40
                        time:   [442.02 ns 449.97 ns 459.55 ns]
                        change: [+1.0719% +4.2921% +7.7698%] (p = 0.01 < 0.05)
                        Performance has regressed.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) high mild
  10 (10.00%) high severe
```

I understand it does more work now but perhaps there are some low-hanging optimisations? E.g. now that it has to use internal iterator, maybe it needs to `flat_map` it into the top iterator instead of just doing `.find` or something like that.

@thomcc Do you want to look into it?

**Comment by RReverser** [comment]
Worst-case scenario, maybe we can only optimise ASCII lowercase strings and forward everything else to owned version... but if possible to optimise unicode scenario too, that would be better.

**Comment by thomcc** [comment]
How much complexity are you willing to accept?

**Comment by RReverser** [comment]
Yeah I thought about using `unic_ucd` which has this [built-in](https://docs.rs/unic-ucd/latest/unic_ucd/fn.changes_when_lowercased.html) but I suspect that will blow up the size of crate considerably.

I'd rather keep it as small and simple as possible.

OTOH how large is the list of exceptions that don't fit into the previous `!is_uppercase` check? I'd be willing to accept build-time complexity if runtime / size cost is not too large.

**Comment by thomcc** [comment]
The size cost should be relatively small. I'll look into it when I have more time.

**Comment by eeeebbbbrrrr** [comment]
> Cow version is now 3x slower than regular one whereas w/o this change it's 2x faster:

I was wondering about this too, but didn't have a chance to bring it up...

Thom and I work together, and one of the projects we're working on really wants to continue to use `cow-utils`, so @RReverser, if you're okay with a bit more complexity and paying a bit of size costs (it sounds like you are?), then we'll definitely get you a PR.

It'll be nice to have the best of all three of Speed, Cost, Correctness for once!

**Comment by RReverser** [comment]
Yeah, as I said, as long as it doesn't blow up the size, it should be fine. It's unfortunate that essentially every crate (std, regex, us) has to ship its own version of unicode data though.

**Pull request merged**